### ISSUE TITLE
Is it possible to display a custom segment of length zero and empty text?

### ISSUE THREAD

**Issue by drinkitup777** [FXMisc/RichTextFX#957]
I'm trying to display an inline image above my user-editable markdown text. This is how I want it to display:

![image 1](https://i.postimg.cc/DZkMFsLv/demo.png)

This is the approach I'm taking to accomplish this:

![image 2](https://i.postimg.cc/NFB3RVs2/demo2.png)

The problem is that the only way I can get my custom `VBox` segment to show up is to make it still be "part of the text"; in other words, make sure `SegmentOps.length(vboxSeg)` returns `> 0`, and make sure `SegmentOps.getText(vboxSeg)` returns a non-empty string.

But doing this makes it selectable and deletable in `GenericStyledArea`. My users will wonder why selecting and copying text containing `vboxSeg` has extra characters in it, and why it takes more than one arrow key press to skip past the `vboxSeg`, when it's only there to display the image in the `TextExt`. Making `vboxSeg` "part of the text" also messes with the API. If I call `GenericStyledArea.getText()`, the string returned by `SegmentOps.getText(vboxSeg)` will also be inside, polluting the returned string.

A possible solution would be to make `vboxSeg` a "phantom" segment, for whom `SegmentOps.length(vboxSeg)` returns `0` and `SegmentOps.getText(vboxSeg)` returns `""`. But doing this seems to make the segment disappear entirely.

Is it possible to have a custom segment of zero length and `""` text still be displayed?

**Comment by Jugen** [comment]
Sorry, I don't think so. If I remember correctly the length has to be at least one so that cursor navigation can work.

**Comment by drinkitup777** [comment]
Hmm... what if I called `GenericStyledArea.getChildren().add(imagePreviewNodeHere)` and override `GenericStyledArea.layoutChildren()` to position it? Would there be any side effects to doing that?

**Comment by Jugen** [comment]
Could work? You'll have to try it out and see. Please give feedback either way, if you do.

**Comment by drinkitup777** [comment]
Nope, this idea's a dud too. First, the `TextExt`s are updated in the view faster than the image preview nodes, so they appear to the user as lagging behind about 1/4 a second or so. Second, because the text flow is virtual, the `TextExt` the image preview node is supposed to be anchored to becomes unpositioned as soon as the `TextExt` is off the screen, even if the image preview node is should still be shown (scrolling up for example, the `TextExt` disappears under the the bottom of the text area, but the image preview node above it should still be shown. without an "anchor", there is no way to position the image preview). Third, translating node positions using `localToParentBounds()` has a ~5 pixel error margin to it, making the image preview node appear to jiggle as the text area is scrolled.

I still have one more idea to make this work. If you want to hear it I'll post the results.

**Comment by Jugen** [comment]
Thanks for the feedback. Yes please post results of your idea ....

**Comment by drinkitup777** [comment]
I think I almost have a working hack, but there's a problem.

The main idea is to simulate an editable text node by listening to mouse events in the custom node and selection changes in `GenericStyledArea`. Here's the custom node:

![demo.png](https://i.postimg.cc/JhXPDfYT/demo.png)

I subscribe to changes of `GenericStyledArea.selectionProperty()` and in the handler, update the pseudo-selection in the custom node. I listen to mouse events (`TextExt.setOnMousePressed()` etc.) on the custom node to update the `GenericStyledArea`'s selection.

But I've run into a SNAFU: when the custom node is clicked, the click handler I've attached with `setOnMousePressed()` fires first, for which the handler calls `GenericStyledArea.selectRange(anchor, caret)`. This is all well and good, but then `GenericStyledArea` runs its default behavior when a custom node is clicked, which sets the selection to `customNodePos + 1`, "overwriting" the selection change done in the `setOnMousePressed()` handler.

Is there any way to prevent `GenericStyledArea` from running its default selection-setting behavior when a custom node is clicked? Some way to ignore mouse events on custom nodes?

**Comment by Jugen** [comment]
Are you consuming the MouseEvent in `setOnMousePressed()` ?  If you are then try changing it to `addEventFilter( MouseEvent.MOUSE_PRESSED, ... )` or maybe MouseEvent.ANY and consume it.
